tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -470,6 +623,9 @@ void CGMSHLSLRuntime::CheckParameterAnnotation(
       Diags.getCustomDiagID(DiagnosticsEngine::Error, "invalid semantic '%0' for %1 %2.%3");
     Diags.Report(SLoc, DiagID) << semName << SM->GetKindName() << SM->GetMajor() << SM->GetMinor();
   }
+
+  if (pSemantic->GetName())
+    DiagnoseSemanticType(SLoc, pSemantic, paramTy, semIndex);
```

Comment:
I just realized a serious limitation of placing all this validation under this codepath.  It only deals with top-level parameter declarations, so it only provides some early rough error checking for a limited set of cases.  To actually check these correctly, you need to generate the final signature shapes.  For that, you have to drill into certain things (vertex arrays, patches, streams), flatten the structures, and apply the semantic inheritance as you do.
This is already done later in the pipeline, but original source locations and clang diagnostics are no longer available at that point.
This is why most semantic validation is performed in HLSignatureLower, where it has the knowledge it needs to do so. 
 I actually think this additional validation needs to be moved there.  Ideally, we could better preserve source locations to improve the diagnostics at that point as well.

Reply:
Moved the type checking logic to `HLSignatureLower`.